### PR TITLE
feat(hydroflow_plus): add prototype of tagging algebraic properties

### PR DESCRIPTION
**Pull request by conor-23**

### CODE REVIEW

File: lattices/src/algebra.rs [hunk 1]
```diff
@@ -221,185 +252,261 @@ pub fn identity<S: Debug + PartialEq + Clone, const N: usize>(
...
+// static SINGLE_FUNCTION_PROPERTIES: [(&str, fn(&[S; N], impl Fn(S, S) -> S)); 6] = [
+//     ("associativity", associativity),
+//     ("commutativity", commutativity),
+//     ("idempotency", idempotency),
+//     ("identity", identity),
+//     ("inverse", inverse),
+//     ("absorbing_element", absorbing_element)];
+
+//Loop through each property in SINGLE_FUNCTION_PROPERTIES and test for them
+pub fn get_single_function_properties<S: Debug + PartialEq + Clone, const N: usize>(
```
Comment (MingweiSamuel) [inline lattices/src/algebra.rs:324]: Use triple slash `///` for doc comments

Comment (MingweiSamuel) [reply lattices/src/algebra.rs:324]: Will also need to go through and fix the other clippy issues - I added a file `precheck.bash` which will generate these errors before waiting for the CI to tell you

File: hydroflow_plus/src/properties.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+use crate::ir::{HfPlusLeaf, HfPlusNode, SeenTees};
+
+// schema : table per algebraic property with the list of expressions that satisfy the property
+
+// interface: can "tag" an expression with a property and it will add it to that table
+// can also run a check to see if an expression satisfies a property
+#[derive(Default)]
+pub struct PropertyDatabase {
+    //
```
Comment (MingweiSamuel) [inline hydroflow_plus/src/properties.rs:13]: remove

Comment (MingweiSamuel) [inline hydroflow_plus/src/properties.rs:13]: Remove empty comment

File: hydroflow_plus/src/properties.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
+use std::collections::HashSet;
+
+use stageleft::*;
+
+use crate::ir::{HfPlusLeaf, HfPlusNode, SeenTees};
+
+// schema : table per algebraic property with the list of expressions that satisfy the property
+
+// interface: can "tag" an expression with a property and it will add it to that table
+// can also run a check to see if an expression satisfies a property
```
Comment (MingweiSamuel) [inline hydroflow_plus/src/properties.rs:10]: nit: make these triple slashes for docs `///`

File: hydroflow_plus/src/properties.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#[derive(Default)]
+pub struct PropertyDatabase {
+    //
+    commutative: HashSet<syn::Expr>,
+}
+
+#[allow(dead_code)]
+// allows us to convert the hydroflow datatype for folds to a binary operation for the algebra property tests
```
Comment (MingweiSamuel) [inline hydroflow_plus/src/properties.rs:18]:
```rust
/// allows us to convert the hydroflow datatype for folds to a binary operation for the algebra property tests
#[allow(dead_code)]
```

Comment (MingweiSamuel) [reply hydroflow_plus/src/properties.rs:18]: Where is this actually used?

File: hydroflow_plus_test/examples/map_reduce.rs [hunk 5]
```diff
@@ -68,11 +71,16 @@ async fn main() {
...
+    // let mut deployment = deployment.into_inner();
 
-    deployment.start().await.unwrap();
+    // deployment.deploy().await.unwrap();
 
-    tokio::signal::ctrl_c().await.unwrap()
+    // deployment.start().await.unwrap();
+
+    // tokio::signal::ctrl_c().await.unwrap()
 }
```
Comment (MingweiSamuel) [inline hydroflow_plus_test/examples/map_reduce.rs:86]: commented-out code?

File: lattices/src/algebra.rs [hunk 6]
```diff
@@ -253,55 +289,120 @@ pub fn identity<S: Debug + PartialEq + Clone, const N: usize>(
...
     f: impl Fn(S, S) -> S,
-) {
+) -> Result<(), &'static str> {
     for [a, b, c] in cartesian_power(items) {
-        assert_eq!(
-            f(a.clone(), f(b.clone(), c.clone())), // f(a, f(b,c)) ie a + (b + c)
-            f(f(a.clone(), b.clone()), c.clone())  // f(f(a,b),c) ie (a + b) + c
-        );
+        if f(a.clone(), f(b.clone(), c.clone())) != // f(a, f(b,c)) ie a + (b + c)
+            f(f(a.clone(), b.clone()), c.clone())
```
Comment (MingweiSamuel) [inline lattices/src/algebra.rs:313]: Seems like one of the comments got lost, good to have both or neither

File: hydroflow_plus/src/properties.rs (hunk 4 above)
Comment (conor-23) [reply hydroflow_plus/src/properties.rs:18]: we don't use it yet, we put it in because we are going to need it when we put in algebraic property checks.

**Review by shadaj — APPROVED** [review]

**Pull request merged**